### PR TITLE
Path Generator

### PR DESCRIPTION
**Pull request by michbr**

### CODE REVIEW

File: Assets/Scripts/Attributes/Endeavour/AttackEnemyBaseAction.cs [hunk 1]
```diff
@@ -92,29 +92,13 @@ public class AttackEnemyBaseAction : Endeavour {
...
-//				if(routePoints[i] == null) {
-//					Debug.LogWarning("Robot '"+controller.name+"' has detected a missing patrol route point!!!");
-//					continue;
-//				}
-//				pathLength += Vector3.Distance(routePoints[i].label.transform.position, routePoints[i+1].label.transform.position);
-//			}
-
-		return getPathLength() * (1f - ((float)nodeIndex/(float)routePoints.Count));
+		getPathLength ();
+		return route.distances [nodeIndex];
```
Comment (semimono) [inline Assets/Scripts/Attributes/Endeavour/AttackEnemyBaseAction.cs]: Is the first line (95) pointless? (punnin' it up)

File: Assets/Scripts/Editor/PathGeneratorGUI.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+		Transform parent = createParentNode();
+		addNextNode (path[0], parent);
+		for (int i = 1; i < path.Count; ++i) {
+			Vector3 step = path[i];
+			addNextNode (step, parent);
+		}
+	}
+
+	private void eliminateInsignificantNodes(List<Vector3> path) {
```
Comment (semimono) [inline Assets/Scripts/Editor/PathGeneratorGUI.cs]: It seems like this method should exist in the PathGenerator and just be referenced here. It's too much logic/algorithm to live in a GUI class.

File: Assets/Scripts/GameControl/GlobalConfig.cs [hunk 3]
```diff
@@ -52,29 +52,15 @@ public class GlobalConfig : NetworkBehaviour {
...
-
-    [Server]
-    public void loseGame() {
-        //NetworkController.networkController.serverClearPlayers();
-        RpcLoseGame();
-    }
+	[Server]
+	public void endGame() {
+		RpcEndGame();
+	}
```
Comment (semimono) [inline Assets/Scripts/GameControl/GlobalConfig.cs]: Awesome to get some of this code out of here. It really belonged more with game mode stuff.

File: Assets/Scripts/Pathing/CubeService.cs [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+		public CreateCubeJob(string key, Vector3 position, Material material, CubeService parent) : base(parent) {
+			this.key = key;
+			this.position = position;
+			this.material = material;
+		}
+		
+		public override void run() {
+			GameObject cube;
+			parent.cubes[key] = cube = GameObject.CreatePrimitive(PrimitiveType.Cube);
```
Comment (semimono) [inline Assets/Scripts/Pathing/CubeService.cs]: I think gizmos would be a lot better, unless for some weird reason we need to be able to see them in game on the regular camera? Can be addressed later.

File: Assets/Scripts/Pathing/BreadthFirstSearch.cs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+					end = new BFSNode(distance, node, current);
+					break;
+					//if we find the end point, stop
+				}
+				if (!found.Contains(getNodeString(node))) {
+					BFSNode bfsNode = new BFSNode(distance, node, current);
+					addVoxelNodeToPathData(bfsNode, distance);
+					found.Add(getNodeString(node));
+					frontier.Enqueue(bfsNode);
+				} else if (distance < nodeDistances[getNodeString(node)].getRank()) {
```
Comment (semimono) [inline Assets/Scripts/Pathing/BreadthFirstSearch.cs]: If we're doing breadth first, how would we ever find a shorter route later? Is this branch ever executed?

File: Assets/Scripts/Pathing/BreadthFirstSearch.cs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+					break;
+					//if we find the end point, stop
+				}
+				if (!found.Contains(getNodeString(node))) {
+					BFSNode bfsNode = new BFSNode(distance, node, current);
+					addVoxelNodeToPathData(bfsNode, distance);
+					found.Add(getNodeString(node));
+					frontier.Enqueue(bfsNode);
+					cubeService.createCube(getNodeString(node), node.getWorldPosition(), Resources.Load<Material>("Materials/Blue"));
+				} else if (distance < nodeDistances[getNodeString(node)].getRank()) {
```
Comment (semimono) [inline Assets/Scripts/Pathing/BreadthFirstSearch.cs]: Above comment here too. Also, do we really need to duplicate the code this much?

File: Assets/Scripts/GameControl/NetworkController.cs [hunk 7]
```diff
@@ -189,9 +190,10 @@ public class NetworkController : MonoBehaviour, SceneLoadListener {
 
 	    string message = netMsg.reader.ReadString();
 	    ClientType type = (ClientType)Enum.Parse(typeof(ClientType), message.Substring(0, 1));
-	    string username = message.Remove(0, 1);
+        int team = Int32.Parse(message.Substring(1, 1));
+	    string username = message.Remove(0, 2);
```
[the remark this answers is not in the document]
Comment (michbr) [reply Assets/Scripts/GameControl/NetworkController.cs]: Agreed. This is pretty delicate. At some point I figure we'll serialize a struct and send it across, but that's going to increase our network traffic significantly.

File: Assets/Scripts/GameControl/TowerBaseSpawner.cs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+	private void OnTriggerEnter(Collider other) {
+		if (towerBase != null)
+			return;
+		RobotController robotController = other.transform.root.GetComponent<RobotController>();
+		if (robotController != null) {
+			int team = robotController.GetComponent<TeamId>().id;
+			if (GlobalConfig.globalConfig.teamGameMode.getJoinedPlayerCount(team) == 0) {
+				Collider[] objects = Physics.OverlapSphere(transform.position, GetComponent<SphereCollider>().radius);
+				foreach (Collider item in objects) {
```
[the remark this answers is not in the document]
Comment (michbr) [reply Assets/Scripts/GameControl/TowerBaseSpawner.cs]: Yes. They can.

File: Assets/Scripts/Attributes/Endeavour/AttackEnemyBaseAction.cs (hunk 1 above)
Comment (michbr) [reply Assets/Scripts/Attributes/Endeavour/AttackEnemyBaseAction.cs]: Potentially not... (Though it does show off my poor design)

**Pull request merged**